{"cells":[{"cell_type":"code","execution_count":1,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":51633,"status":"ok","timestamp":1689032480577,"user":{"displayName":"이동원","userId":"05424808703187982988"},"user_tz":-540},"id":"Yw71Zr6ehe7y","outputId":"b261b025-060d-4097-8915-6262984303a4"},"outputs":[{"name":"stdout","output_type":"stream","text":["Mounted at /content/drive\n"]}],"source":["# This mounts your Google Drive to the Colab VM.\n","from google.colab import drive\n","drive.mount('/content/drive')"]},{"cell_type":"code","execution_count":2,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":762,"status":"ok","timestamp":1689032525518,"user":{"displayName":"이동원","userId":"05424808703187982988"},"user_tz":-540},"id":"_tMvx2Qvh70b","outputId":"147a5b91-458f-4e81-b4e0-4b2659df5460"},"outputs":[{"name":"stdout","output_type":"stream","text":["/content/drive/.shortcut-targets-by-id/1ZfbeRJGYGC-9d0cU_aG1Cprzhc40b-7T/analytics/결과\n"]}],"source":["FOLDERNAME = 'Colab Notebooks/analytics/결과'\n","assert FOLDERNAME is not None, \"[!] Enter the foldername.\"\n","\n","import sys\n","sys.path.append('/content/drive/MyDrive/{}'.format(FOLDERNAME))\n","\n","# Change dariectory to current folder\n","%cd /content/drive/MyDrive/$FOLDERNAME"]},{"cell_type":"code","execution_count":3,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":4496,"status":"ok","timestamp":1689032535264,"user":{"displayName":"이동원","userId":"05424808703187982988"},"user_tz":-540},"id":"4eoGnZwSjpnG","outputId":"0dfcccbd-cc9a-4446-f769-f1c6dcf6fa91"},"outputs":[{"name":"stderr","output_type":"stream","text":["/usr/local/lib/python3.10/dist-packages/jaxlib/xla_client.py:225: DeprecationWarning: ml_dtypes.float8_e4m3b11 is deprecated. Use ml_dtypes.float8_e4m3b11fnuz\n","  float8_e4m3b11fnuz = ml_dtypes.float8_e4m3b11\n"]}],"source":["import gym\n","import tensorflow as tf\n","from tensorflow.keras import layers\n","import numpy as np\n","import matplotlib.pyplot as plt\n","tf.config.run_functions_eagerly(True)"]},{"cell_type":"code","execution_count":4,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":372,"status":"ok","timestamp":1689032544947,"user":{"displayName":"이동원","userId":"05424808703187982988"},"user_tz":-540},"id":"-ldl7IVTjv-Y","outputId":"a466af2f-5344-48de-ee15-5e2b7251d1b0"},"outputs":[{"name":"stdout","output_type":"stream","text":["Size of State Space -\u003e  3\n","Size of Action Space -\u003e  1\n","Max Value of Action -\u003e  2.0\n","Min Value of Action -\u003e  -2.0\n"]},{"name":"stderr","output_type":"stream","text":["/usr/local/lib/python3.10/dist-packages/gym/core.py:317: DeprecationWarning: \u001b[33mWARN: Initializing wrapper in old step API which returns one bool instead of two. It is recommended to set `new_step_api=True` to use new step API. This will be the default behaviour in future.\u001b[0m\n","  deprecation(\n","/usr/local/lib/python3.10/dist-packages/gym/wrappers/step_api_compatibility.py:39: DeprecationWarning: \u001b[33mWARN: Initializing environment in old step API which returns one bool instead of two. It is recommended to set `new_step_api=True` to use new step API. This will be the default behaviour in future.\u001b[0m\n","  deprecation(\n"]}],"source":["problem = \"Pendulum-v1\"\n","env = gym.make(problem)\n","\n","num_states = env.observation_space.shape[0]\n","print(\"Size of State Space -\u003e  {}\".format(num_states))\n","num_actions = env.action_space.shape[0]\n","print(\"Size of Action Space -\u003e  {}\".format(num_actions))\n","\n","upper_bound = env.action_space.high[0]\n","lower_bound = env.action_space.low[0]\n","\n","print(\"Max Value of Action -\u003e  {}\".format(upper_bound))\n","print(\"Min Value of Action -\u003e  {}\".format(lower_bound))\n","\n","critic_value_list=[]"]},{"cell_type":"code","execution_count":5,"metadata":{"executionInfo":{"elapsed":6,"status":"ok","timestamp":1689032556224,"user":{"displayName":"이동원","userId":"05424808703187982988"},"user_tz":-540},"id":"EO6V2Ed_j2ug"},"outputs":[],"source":["class OUActionNoise:\n","    def __init__(self, mean, std_deviation, theta=0.15, dt=1e-2, x_initial=None):\n","        self.theta = theta\n","        self.mean = mean\n","        self.std_dev = std_deviation\n","        self.dt = dt\n","        self.x_initial = x_initial\n","        self.reset()\n","\n","    def __call__(self):\n","        # Formula taken from https://www.wikipedia.org/wiki/Ornstein-Uhlenbeck_process.\n","        x = (\n","            self.x_prev\n","            + self.theta * (self.mean - self.x_prev) * self.dt\n","            + self.std_dev * np.sqrt(self.dt) * np.random.normal(size=self.mean.shape)\n","        )\n","        # Store x into x_prev\n","        # Makes next noise dependent on current one\n","        self.x_prev = x\n","        return x\n","\n","    def reset(self):\n","        if self.x_initial is not None:\n","            self.x_prev = self.x_initial\n","        else:\n","            self.x_prev = np.zeros_like(self.mean)"]},{"cell_type":"code","execution_count":6,"metadata":{"executionInfo":{"elapsed":478,"status":"ok","timestamp":1689032561716,"user":{"displayName":"이동원","userId":"05424808703187982988"},"user_tz":-540},"id":"yv1p11Vwj5_5"},"outputs":[],"source":["class Buffer:\n","    def __init__(self, buffer_capacity=100000, batch_size=64):\n","        # Number of \"experiences\" to store at max\n","        self.buffer_capacity = buffer_capacity\n","        # Num of tuples to train on.\n","        self.batch_size = batch_size\n","\n","        # Its tells us num of times record() was called.\n","        self.buffer_counter = 0\n","\n","        # Instead of list of tuples as the exp.replay concept go\n","        # We use different np.arrays for each tuple element\n","        self.state_buffer = np.zeros((self.buffer_capacity, num_states))\n","        self.action_buffer = np.zeros((self.buffer_capacity, num_actions))\n","        self.reward_buffer = np.zeros((self.buffer_capacity, 1))\n","        self.next_state_buffer = np.zeros((self.buffer_capacity, num_states))\n","\n","    # Takes (s,a,r,s') obervation tuple as input\n","    def record(self, obs_tuple):\n","        # Set index to zero if buffer_capacity is exceeded,\n","        # replacing old records\n","        index = self.buffer_counter % self.buffer_capacity\n","\n","        self.state_buffer[index] = obs_tuple[0]\n","        self.action_buffer[index] = obs_tuple[1]\n","        self.reward_buffer[index] = obs_tuple[2]\n","        self.next_state_buffer[index] = obs_tuple[3]\n","\n","        self.buffer_counter += 1\n","\n","    # Eager execution is turned on by default in TensorFlow 2. Decorating with tf.function allows\n","    # TensorFlow to build a static graph out of the logic and computations in our function.\n","    # This provides a large speed up for blocks of code that contain many small TensorFlow operations such as this one.\n","    @tf.function\n","    def update(\n","        self, state_batch, action_batch, reward_batch, next_state_batch,\n","    ):\n","        # Training and updating Actor \u0026 Critic networks.\n","        # See Pseudo Code.\n","        with tf.GradientTape() as tape:\n","            target_actions = target_actor(next_state_batch, training=True)\n","            y = reward_batch + gamma * target_critic(\n","                [next_state_batch, target_actions], training=True\n","            )\n","            critic_value = critic_model([state_batch, action_batch], training=True)\n","            critic_loss = tf.math.reduce_mean(tf.math.square(y - critic_value))\n","\n","        critic_grad = tape.gradient(critic_loss, critic_model.trainable_variables)\n","        critic_optimizer.apply_gradients(\n","            zip(critic_grad, critic_model.trainable_variables)\n","        )\n","\n","        with tf.GradientTape() as tape:\n","            actions = actor_model(state_batch, training=True)\n","            critic_value = critic_model([state_batch, actions], training=True)\n","\n","            # Used `-value` as we want to maximize the value given\n","            # by the critic for our actions\n","            actor_loss = -tf.math.reduce_mean(critic_value)\n","\n","        actor_grad = tape.gradient(actor_loss, actor_model.trainable_variables)\n","        actor_optimizer.apply_gradients(\n","            zip(actor_grad, actor_model.trainable_variables)\n","        )\n","\n","    # We compute the loss and update parameters\n","    def learn(self):\n","        # Get sampling range\n","        record_range = min(self.buffer_counter, self.buffer_capacity)\n","        # Randomly sample indices\n","        batch_indices = np.random.choice(record_range, self.batch_size)\n","\n","        # Convert to tensors\n","        critic_value_list.append(self.state_buffer[batch_indices])\n","        state_batch = tf.convert_to_tensor(self.state_buffer[batch_indices])\n","        action_batch = tf.convert_to_tensor(self.action_buffer[batch_indices])\n","        reward_batch = tf.convert_to_tensor(self.reward_buffer[batch_indices])\n","        reward_batch = tf.cast(reward_batch, dtype=tf.float32)\n","        next_state_batch = tf.convert_to_tensor(self.next_state_buffer[batch_indices])\n","\n","        self.update(state_batch, action_batch, reward_batch, next_state_batch)\n","\n","\n","# This update target parameters slowly\n","# Based on rate `tau`, which is much less than one.\n","@tf.function\n","def update_target(target_weights, weights, tau):\n","    for (a, b) in zip(target_weights, weights):\n","        a.assign(b * tau + a * (1 - tau))"]},{"cell_type":"code","execution_count":7,"metadata":{"executionInfo":{"elapsed":365,"status":"ok","timestamp":1689032581381,"user":{"displayName":"이동원","userId":"05424808703187982988"},"user_tz":-540},"id":"YJdZQPtMj8es"},"outputs":[],"source":["def get_actor():\n","    # Initialize weights between -3e-3 and 3-e3\n","    last_init = tf.random_uniform_initializer(minval=-0.003, maxval=0.003)\n","\n","    inputs = layers.Input(shape=(num_states,))\n","    out = layers.Dense(256, activation=\"relu\")(inputs)\n","    out = layers.Dense(256, activation=\"relu\")(out)\n","    outputs = layers.Dense(1, activation=\"tanh\", kernel_initializer=last_init)(out)\n","\n","    # Our upper bound is 2.0 for Pendulum.\n","    outputs = outputs * upper_bound\n","    model = tf.keras.Model(inputs, outputs)\n","    return model\n","\n","\n","def get_critic():\n","    # State as input\n","    state_input = layers.Input(shape=(num_states))\n","    state_out = layers.Dense(16, activation=\"relu\")(state_input)\n","    state_out = layers.Dense(32, activation=\"relu\")(state_out)\n","\n","    # Action as input\n","    action_input = layers.Input(shape=(num_actions))\n","    action_out = layers.Dense(32, activation=\"relu\")(action_input)\n","\n","    # Both are passed through seperate layer before concatenating\n","    concat = layers.Concatenate()([state_out, action_out])\n","\n","    out = layers.Dense(256, activation=\"relu\")(concat)\n","    out = layers.Dense(256, activation=\"relu\")(out)\n","    outputs = layers.Dense(1)(out)\n","\n","    # Outputs single value for give state-action\n","    model = tf.keras.Model([state_input, action_input], outputs)\n","\n","    return model"]},{"cell_type":"code","execution_count":8,"metadata":{"executionInfo":{"elapsed":427,"status":"ok","timestamp":1689032589395,"user":{"displayName":"이동원","userId":"05424808703187982988"},"user_tz":-540},"id":"u5qzxzvkj_lr"},"outputs":[],"source":["def policy(state, noise_object):\n","    sampled_actions = tf.squeeze(actor_model(state))\n","    noise = noise_object()\n","    # Adding noise to action\n","    sampled_actions = sampled_actions.numpy() + noise\n","\n","    # We make sure action is within bounds\n","    legal_action = np.clip(sampled_actions, lower_bound, upper_bound)\n","\n","    return [np.squeeze(legal_action)]"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"background_save":true,"base_uri":"https://localhost:8080/"},"id":"69gutxZZkFl1"},"outputs":[{"name":"stderr","output_type":"stream","text":[" 35%|███▌      | 7/20 [1:59:41\u003c3:42:16, 1025.90s/it]\n","ERROR:root:Internal Python error in the inspect module.\n","Below is the traceback from this internal error.\n","\n","ERROR:root:Internal Python error in the inspect module.\n","Below is the traceback from this internal error.\n","\n","ERROR:root:Internal Python error in the inspect module.\n","Below is the traceback from this internal error.\n","\n"]},{"name":"stdout","output_type":"stream","text":["OSError: [Errno 107] Transport endpoint is not connected\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3553, in run_code\n","    exec(code_obj, self.user_global_ns, self.user_ns)\n","  File \"\u003cipython-input-10-4dd737517e51\u003e\", line 95, in \u003ccell line: 94\u003e\n","    with open(f'list_to_csv_frozen-{number_of_episode}{indexing_str}.csv','a',newline='') as f:\n","OSError: [Errno 107] Transport endpoint is not connected\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2099, in showtraceback\n","    stb = value._render_traceback_()\n","AttributeError: 'OSError' object has no attribute '_render_traceback_'\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1101, in get_records\n","    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 248, in wrapped\n","    return f(*args, **kwargs)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n","    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n","  File \"/usr/lib/python3.10/inspect.py\", line 1662, in getinnerframes\n","    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n","  File \"/usr/lib/python3.10/inspect.py\", line 1620, in getframeinfo\n","    filename = getsourcefile(frame) or getfile(frame)\n","  File \"/usr/lib/python3.10/inspect.py\", line 829, in getsourcefile\n","    module = getmodule(object, filename)\n","  File \"/usr/lib/python3.10/inspect.py\", line 861, in getmodule\n","    file = getabsfile(object, _filename)\n","  File \"/usr/lib/python3.10/inspect.py\", line 845, in getabsfile\n","    return os.path.normcase(os.path.abspath(_filename))\n","  File \"/usr/lib/python3.10/posixpath.py\", line 384, in abspath\n","    cwd = os.getcwd()\n","OSError: [Errno 107] Transport endpoint is not connected\n","OSError: [Errno 107] Transport endpoint is not connected\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3553, in run_code\n","    exec(code_obj, self.user_global_ns, self.user_ns)\n","  File \"\u003cipython-input-10-4dd737517e51\u003e\", line 95, in \u003ccell line: 94\u003e\n","    with open(f'list_to_csv_frozen-{number_of_episode}{indexing_str}.csv','a',newline='') as f:\n","OSError: [Errno 107] Transport endpoint is not connected\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2099, in showtraceback\n","    stb = value._render_traceback_()\n","AttributeError: 'OSError' object has no attribute '_render_traceback_'\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3473, in run_ast_nodes\n","    if (await self.run_code(code, result,  async_=asy)):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3575, in run_code\n","    self.showtraceback(running_compiled_code=True)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2101, in showtraceback\n","    stb = self.InteractiveTB.structured_traceback(etype,\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1367, in structured_traceback\n","    return FormattedTB.structured_traceback(\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1267, in structured_traceback\n","    return VerboseTB.structured_traceback(\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1124, in structured_traceback\n","    formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n","    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n","    return len(records), 0\n","TypeError: object of type 'NoneType' has no len()\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2099, in showtraceback\n","    stb = value._render_traceback_()\n","AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1101, in get_records\n","    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 248, in wrapped\n","    return f(*args, **kwargs)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n","    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n","  File \"/usr/lib/python3.10/inspect.py\", line 1662, in getinnerframes\n","    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n","  File \"/usr/lib/python3.10/inspect.py\", line 1620, in getframeinfo\n","    filename = getsourcefile(frame) or getfile(frame)\n","  File \"/usr/lib/python3.10/inspect.py\", line 829, in getsourcefile\n","    module = getmodule(object, filename)\n","  File \"/usr/lib/python3.10/inspect.py\", line 861, in getmodule\n","    file = getabsfile(object, _filename)\n","  File \"/usr/lib/python3.10/inspect.py\", line 845, in getabsfile\n","    return os.path.normcase(os.path.abspath(_filename))\n","  File \"/usr/lib/python3.10/posixpath.py\", line 384, in abspath\n","    cwd = os.getcwd()\n","OSError: [Errno 107] Transport endpoint is not connected\n","OSError: [Errno 107] Transport endpoint is not connected\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3553, in run_code\n","    exec(code_obj, self.user_global_ns, self.user_ns)\n","  File \"\u003cipython-input-10-4dd737517e51\u003e\", line 95, in \u003ccell line: 94\u003e\n","    with open(f'list_to_csv_frozen-{number_of_episode}{indexing_str}.csv','a',newline='') as f:\n","OSError: [Errno 107] Transport endpoint is not connected\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2099, in showtraceback\n","    stb = value._render_traceback_()\n","AttributeError: 'OSError' object has no attribute '_render_traceback_'\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3473, in run_ast_nodes\n","    if (await self.run_code(code, result,  async_=asy)):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3575, in run_code\n","    self.showtraceback(running_compiled_code=True)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2101, in showtraceback\n","    stb = self.InteractiveTB.structured_traceback(etype,\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1367, in structured_traceback\n","    return FormattedTB.structured_traceback(\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1267, in structured_traceback\n","    return VerboseTB.structured_traceback(\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1124, in structured_traceback\n","    formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n","    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n","    return len(records), 0\n","TypeError: object of type 'NoneType' has no len()\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2099, in showtraceback\n","    stb = value._render_traceback_()\n","AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3030, in _run_cell\n","    return runner(coro)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/async_helpers.py\", line 78, in _pseudo_sync_runner\n","    coro.send(None)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3257, in run_cell_async\n","    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3492, in run_ast_nodes\n","    self.showtraceback()\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2101, in showtraceback\n","    stb = self.InteractiveTB.structured_traceback(etype,\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1367, in structured_traceback\n","    return FormattedTB.structured_traceback(\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1267, in structured_traceback\n","    return VerboseTB.structured_traceback(\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1142, in structured_traceback\n","    formatted_exceptions += self.format_exception_as_a_whole(etype, evalue, etb, lines_of_context,\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n","    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n","    return len(records), 0\n","TypeError: object of type 'NoneType' has no len()\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2099, in showtraceback\n","    stb = value._render_traceback_()\n","AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1101, in get_records\n","    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 248, in wrapped\n","    return f(*args, **kwargs)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n","    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n","  File \"/usr/lib/python3.10/inspect.py\", line 1662, in getinnerframes\n","    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n","  File \"/usr/lib/python3.10/inspect.py\", line 1620, in getframeinfo\n","    filename = getsourcefile(frame) or getfile(frame)\n","  File \"/usr/lib/python3.10/inspect.py\", line 829, in getsourcefile\n","    module = getmodule(object, filename)\n","  File \"/usr/lib/python3.10/inspect.py\", line 861, in getmodule\n","    file = getabsfile(object, _filename)\n","  File \"/usr/lib/python3.10/inspect.py\", line 845, in getabsfile\n","    return os.path.normcase(os.path.abspath(_filename))\n","  File \"/usr/lib/python3.10/posixpath.py\", line 384, in abspath\n","    cwd = os.getcwd()\n","OSError: [Errno 107] Transport endpoint is not connected\n"]}],"source":["from tqdm import tqdm\n","from datetime import datetime\n","def main(total_episodes):\n","  global actor_model, critic_model, target_actor, target_critic, critic_optimizer\n","  global actor_optimizer, gamma\n","  std_dev = 0.2\n","  ou_noise = OUActionNoise(mean=np.zeros(1), std_deviation=float(std_dev) * np.ones(1))\n","\n","  actor_model = get_actor()\n","  critic_model = get_critic()\n","\n","  target_actor = get_actor()\n","  target_critic = get_critic()\n","\n","  # Making the weights equal initially\n","  target_actor.set_weights(actor_model.get_weights())\n","  target_critic.set_weights(critic_model.get_weights())\n","\n","  # Learning rate for actor-critic models\n","  critic_lr = 0.002\n","  actor_lr = 0.001\n","\n","  critic_optimizer = tf.keras.optimizers.Adam(critic_lr)\n","  actor_optimizer = tf.keras.optimizers.Adam(actor_lr)\n","\n","  # Discount factor for future rewards\n","  gamma = 0.99\n","  # Used to update target networks\n","  tau = 0.005\n","\n","  buffer = Buffer(50000, 64)\n","\n","\n","  # To store reward history of each episode\n","  ep_reward_list = []\n","  # To store time history of each episode\n","  datetime_list=[]\n","  start = datetime.now()\n","\n","  #출력 모아놓는 코드\n","  a=[]\n","\n","  # Takes about 9 min to train\n","  for ep in range(total_episodes):\n","\n","      prev_state = env.reset()\n","      episodic_reward = 0\n","\n","      while True:\n","          # Uncomment this to see the Actor in action\n","          # But not in a python notebook.\n","          # env.render()\n","\n","          tf_prev_state = tf.expand_dims(tf.convert_to_tensor(prev_state), 0)\n","\n","          action = policy(tf_prev_state, ou_noise)\n","          # Recieve state and reward from environment.\n","          state, reward, done, info = env.step(action)\n","\n","          buffer.record((prev_state, action, reward, state))\n","          episodic_reward += reward\n","\n","          buffer.learn()\n","          if ep % 5 == 4:\n","            update_target(target_actor.variables, actor_model.variables, 5 * tau)\n","            update_target(target_critic.variables, critic_model.variables, 5 * tau)\n","\n","          # \"deleted\" basic code\n","\n","          # update_target(target_actor.variables, actor_model.variables, tau)\n","          # update_target(target_critic.variables, critic_model.variables, tau)\n","\n","          # End this episode when `done` is True\n","          if done:\n","              break\n","\n","          prev_state = state\n","\n","      # 끝난 시간 얻기\n","      end = datetime.now()\n","\n","      ep_reward_list.append(episodic_reward)\n","\n","      datetime_list.append(end-start)\n","\n","  return datetime_list, ep_reward_list\n","\n","import csv\n","\n","number_of_episode=50\n","n=20\n","\n","indexing_str=datetime.today().strftime(\"%Y%m%d_%H%M%S\")\n","for _ in tqdm(range(n)):\n","    with open(f'list_to_csv_frozen-{number_of_episode}{indexing_str}.csv','a',newline='') as f:\n","      writer = csv.writer(f)\n","      data_for_save=main(number_of_episode) #totalepisode\n","      writer.writerow(data_for_save[0])\n","      writer.writerow(data_for_save[1])"]}],"metadata":{"colab":{"name":"","provenance":[{"file_id":"1VedkvFwzaU4cs3jK4FRSYSKhZjAUgY9J","timestamp":1688888700729},{"file_id":"1P1SQ9Q1a0olUnMnUpq4Nilk27OJvS9Qf","timestamp":1688888540877}],"version":""},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}